File: algorithms/java/144_binary_tree_preorder_traversal/binary_tree_preorder_traversal.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import tree.BinTree;
import tree.BinTree.TreeNode;

/**
 * @author Peng
 * 
 * Problem Description:
 *      Given a binary tree, return the preorder traversal 
 *      of its nodes' values.(DLR)
 */
public class binary_tree_preorder_traversal {

    /**
     * Solution 1:
     *      using recursive
     * @param root
     * @return nodes' values list of DLR
     */
    public static List<Integer> preorderTraversal_1(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if(root == null)  return res;
        res.add(root.val);
        res.addAll(preorderTraversal_1(root.left));
        res.addAll(preorderTraversal_1(root.right));
        
        return res;
    }
    
    /**
     * Solution 2:
     *      using iteration
     * @param root
     * @return nodes' values list of DLR
     */
    public static List<Integer> preorderTraversal_2(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> st = new Stack<TreeNode>();
        
        while(st.empty() != true || root != null){
            if(root != null) {
                res.add(root.val);
                st.push(root);
                root = root.left;
            }
            else {
                root = st.pop().right;
            }
        }
        
        return res;
    }
    
    
    public static void main(String[] args) {
        // creating a binary tree.
        Integer[] data = {1, 2, 3, null, 5, null, null};
        BinTree bt = new BinTree();
        bt.bTreeRoot = bt.createTree(data, 0, 7);  
        
        List<Integer> list = preorderTraversal_2(bt.bTreeRoot);
        for(Integer i : list) { 
            System.out.printf(i + "  ");
        }
    }

}

File: algorithms/python/144_binary_tree_preorder_traversal/binary_tree_preorder_traversal.py
# coding: utf8
# python 2.7
"""
Link: https://leetcode-cn.com/problems/binary-tree-preorder-traversal
Given a binary tree, return the preorder traversal of its nodes' values.

Example:
    Input: [1,null,2,3]
       1
        \
         2
        /
       3
    Output: [1,2,3]
Follow up: Recursive solution is trivial, could you do it iteratively?
"""
import sys
sys.path.append("../")
from self_made_libs.binary_tree import BTreeHelper, TreeNode

class Solution(object):

    def preorderTraversal_v1(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        :idea: do it iteratively, complexity: time O(n) space O(n)
        """
        res = []
        S = [root] if root else []
        while S:
            p = S.pop()
            res.append(p.val)
            if p.right:
                S.append(p.right)
            if p.left:
                S.append(p.left)
        return res

    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        :idea: do it recursively
        """
        res = []
        if not root:
            return res
        res.append(root.val)
        res.extend(self.preorderTraversal(root.left))
        res.extend(self.preorderTraversal(root.right))
        return res

if __name__ == '__main__':
    BT = BTreeHelper()
    S = Solution()
    arr = [1,2,2,None,3,None,3]
    root = BT.build_a_tree(arr)
    res = S.preorderTraversal_v1(root)
    print res
